### PR TITLE
WIP - Ledger pruning: Proposed Ledger API extensions

### PR DESCRIPTION
**Pull request by oliverse-da**
Meant to get detail discussions going next to high-level design. Only contains ledger API skeleton for now.

Round 1: solicit and incorporate feedback on things like naming, exceptions, boundary conditions (e.g. if pruning point is too recent).

Will add ReadService/WriteService changes shortly also as WIP.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

### CODE REVIEW

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+option csharp_namespace = "Com.DigitalAsset.Ledger.Api.V1.Admin";
+
+// Status: experimental interface, will change before it is deemed production
+// ready
+
+// Prunes/truncates the "oldest" ledger transactions from the ledger (specifically the Ledger Api Server) to
+// 1. keep the "inactive" portion of the ledger to a manageable size (i.e. removing the portion of the ledger that does
+//    not affect the set of allowed, future commands.
+// 2. remove inactive state to honor right-to-forget guarantees.
+service LedgerPruningService {
```
Comment (stefanobaghino-da) [inline ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto]: Would it make sense to have a `pruneByOffset` and a `pruneByTime` method rather than using and relying on unions to discriminate?

Comment (oliverse-da) [reply ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto]: Sure, seems reasonable - will change.

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  // Returns an error if the ledger api server is not configured to support pruning.
+  rpc PruneByOffset (PruneByOffsetRequest) returns (PruneResponse);
+
+  // Prune the ledger specifying the point in time or offset-space before which ledger transactions should be removed.
+  // Returns an error if the ledger api server is not configured to support pruning or is configured with a static
+  // time (such that time is does not differentiate specific pruning points).
+  rpc PruneByTime (PruneByTimeRequest) returns (PruneResponse);
+}
+
+message PruneByOffsetRequest {
```
Comment (stefanobaghino-da) [inline ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto]: Looks like `PruneByOffsetRequest` and `PruneByTimeRequest` have their only member swapped. :wink:

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  google.protobuf.Timestamp prune_before = 1;
+}
+
+message PruneByTimeRequest {
+  string prune_before = 1;
+}
+
+message PruneResponse {
```
Comment (stefanobaghino-da) [inline ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto]: Since we split the method, I'd say we can also split the response.

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  rpc PruneByOffset (PruneByOffsetRequest) returns (PruneResponse);
+
+  // Prune the ledger specifying the point in time or offset-space before which ledger transactions should be removed.
+  // Returns an error if the ledger api server is not configured to support pruning or is configured with a static
+  // time (such that time is does not differentiate specific pruning points).
+  rpc PruneByTime (PruneByTimeRequest) returns (PruneResponse);
+}
+
+message PruneByOffsetRequest {
+  google.protobuf.Timestamp prune_before = 1;
```
Comment (stefanobaghino-da) [inline ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto]: What does `before` mean? Is it inclusive? The name seems to imply an exclusive behavior, which makes sense: it is easier to manage on the `byTime` alternative and it would be nice to have a symmetric meaning on the `byOffset` alternative. Whichever way we want to adopt, it's probably better to document the semantics for clarity's sake.

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+// Status: experimental interface, will change before it is deemed production
+// ready
+
+// Prunes/truncates the "oldest" ledger transactions from the ledger (specifically the Ledger Api Server) by removing
+// a portion of the ledger that does not affect the set of allowed, future commands). This enables:
+// 1. keeping the "inactive" portion of the ledger to a manageable size and
+// 2. removing inactive state to honor right-to-forget guarantees.
+service LedgerPruningService {
+
+  // Prune the ledger specifying the point in time or offset-space before which ledger transactions should be removed.
```
Comment (stefanobaghino-da) [inline ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto]: The comment on both endpoints should be update to reflect the new behavior.

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto (hunk 2 above)
Comment (oliverse-da) [reply ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto]: Thanks - that's what I call an intuitive API

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import "google/protobuf/timestamp.proto";
+
+option java_outer_classname = "LedgerPruningServiceOuterClass";
+option java_package = "com.digitalasset.ledger.api.v1.admin";
+option csharp_namespace = "Com.DigitalAsset.Ledger.Api.V1.Admin";
+
+// Status: experimental interface, will change before it is deemed production
+// ready
+
+// Prunes/truncates the "oldest" ledger transactions from the ledger (specifically the Ledger Api Server) by removing
```
Comment (dajmaki) [inline ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto]: Isn't this service supposed to be about pruning the participant and not the underlying ledger (which of course could be distributed and not prunable from a single point). I would reword the comments to talk explicitly about the participant, though we'll likely need to explain here what this would mean in practice (e.g. the index is pruned, and possible some participant-local storage).

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+// pruning point is
+// 1. in the future or
+// 2. too recent, attempting pruning transactions tracked as inflight-transactions by the CommandService.
+message PruneByOffsetResponse {
+  // Exclusive offset up to which pruning has actually been applied.
+  string pruned_offset = 1;
+}
+
+message PruneByTimeResponse {
+  // Exclusive point in time up to which pruning has actually been applied.
```
Comment (dajmaki) [inline ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/ledger_pruning_service.proto]: we could also return the offset.
